src/devices/video/vrender0.cpp

```diff
@@ -321,202 +322,127 @@ static uint16_t Alpha(QuadInfo *Quad, uint16_t Src, uint16_t Dst)
 		dmulb = 0x100 - dmulb;
 	}
 
-
-	dcr = (scr * smulr + dcr * dmulr) >> 8;
-	if (dcr > 0xff)
-		dcr = 0xff;
-
-	dcg = (scg * smulg + dcg * dmulg) >> 8;
-	if (dcg > 0xff)
-		dcg = 0xff;
-
-	dcb = (scb * smulb + dcb * dmulb) >> 8;
-	if (dcb > 0xff)
-		dcb = 0xff;
+	dcr = m_additive_lookup[m_multiply_lookup[scr][smulr]][m_multiply_lookup[dcr][dmulr]];
+	dcg = m_additive_lookup[m_multiply_lookup[scg][smulg]][m_multiply_lookup[dcg][dmulg]];
+	dcb = m_additive_lookup[m_multiply_lookup[scb][smulb]][m_multiply_lookup[dcb][dmulb]];
 
 	return RGB16(dcr, dcg, dcb);
 }
 
-#define TILENAME(bpp, t, a) \
-static void DrawQuad##bpp##t##a(QuadInfo *Quad)
-
 //TRUST ON THE COMPILER OPTIMIZATIONS
-#define TILETEMPL(bpp, t, a) \
-TILENAME(bpp, t, a)\
-{\
-	uint32_t TransColor = Quad->Trans ? RGB32TO16(Quad->TransColor) : NOTRANSCOLOR;\
-	uint32_t x, y;\
-	uint16_t *line = Quad->Dest;\
-	uint32_t y_tx = Quad->Tx, y_ty = Quad->Ty;\
-	uint32_t x_tx, x_ty;\
-	uint32_t Maskw = Quad->TWidth - 1;\
-	uint32_t Maskh = Quad->THeight - 1;\
-	uint32_t W = Quad->TWidth >> 3;\
-\
-	for (y = 0; y < Quad->h; ++y)\
-	{\
-		uint16_t *pixel = line;\
-		x_tx = y_tx;\
-		x_ty = y_ty;\
-		for (x = 0; x < Quad->w; ++x)\
-		{\
-			uint32_t Offset;\
-			uint32_t tx = x_tx >> 9;\
-			uint32_t ty = x_ty >> 9;\
-			uint16_t Color;\
-			if (Quad->Clamp)\
-			{\
-				if (tx > Maskw)\
-					goto Clamped;\
-				if (ty > Maskh)\
-					goto Clamped;\
-			}\
-			else\
-			{\
-				tx &= Maskw;\
-				ty &= Maskh;\
-			}\
-\
-			if(t)\
-			{\
-				uint32_t Index=Quad->Tile[(ty>>3)*(W)+(tx>>3)];\
-				Offset=(Index<<6)+((ty&7)<<3)+(tx&7);\
-				if(Index==0) goto Clamped;\
-			}\
-			else\
-				Offset = ty * (Quad->TWidth) + tx;\
-\
-			if (bpp == 4)\
-			{\
-				uint8_t Texel = Quad->u.Imageb[Offset / 2];\
-				if (Offset & 1)\
-					Texel &= 0xf;\
-				else\
-					Texel = (Texel >> 4) & 0xf;\
-				Color = Quad->Pal[Texel];\
-			}\
-			else if (bpp == 8)\
-			{\
-				uint8_t Texel = Quad->u.Imageb[Offset];\
-				Color = Quad->Pal[Texel];\
-			}\
-			else if (bpp == 16)\
-			{\
-				Color = Quad->u.Imagew[Offset];\
-			}\
-			if (Color != TransColor)\
-			{\
-				if (a == 1)\
-					*pixel = Alpha(Quad, Color, *pixel);\
-				else if (a == 2)\
-					*pixel = Shade(Color, Quad->Shade);\
-				else\
-					*pixel = Color;\
-			}\
-			Clamped:\
-			++pixel;\
-			x_tx += Quad->Txdx;\
-			x_ty += Quad->Tydx;\
-		}\
-		line += Quad->Pitch;\
-		y_tx += Quad->Txdy;\
-		y_ty += Quad->Tydy;\
-	}\
-}
-TILETEMPL(16,0,0) TILETEMPL(16,0,1) TILETEMPL(16,0,2)
-TILETEMPL(16,1,0) TILETEMPL(16,1,1) TILETEMPL(16,1,2)
-
-TILETEMPL(8,0,0) TILETEMPL(8,0,1) TILETEMPL(8,0,2)
-TILETEMPL(8,1,0) TILETEMPL(8,1,1) TILETEMPL(8,1,2)
-
-TILETEMPL(4,0,0) TILETEMPL(4,0,1) TILETEMPL(4,0,2)
-TILETEMPL(4,1,0) TILETEMPL(4,1,1) TILETEMPL(4,1,2)
-
-#undef TILENAME
-#define TILENAME(bpp, t, a) \
-DrawQuad##bpp##t##a
-
-
-static void DrawQuadFill(QuadInfo *Quad)
+void vr0video_device::DrawQuad(u8 bpp, bool tiled, QuadInfo *Quad)
```

Comment:
In general I'm gonna guess this code is significantly slower. 
Please consider using a functional approach to the critical paths e.g. all of the inner if-else that this function contains.